### PR TITLE
Feat/cut to coast

### PR DESCRIPTION
**Pull request by jacott**
Apply a geojson cutline to images being run through cogify 

#### Source Code Documentation Tasks:
- [ ] README updated (where applicable)
- [ ] CHANGELOG (Unreleased section) updated
- [ ] Docstrings / comments included to help explain code

#### User Documentation Tasks:
- [ ] Confluence user guide updated (where applicable)

#### Testing Tasks:
- [ ] Added tests that fail without this change
- [ ] All tests are passing in development environment
- [ ] Reviewers assigned
- [ ] Linked to main issue for ZenHub board

### CODE REVIEW

File: packages/cog/src/cli/cogify/action.cog.ts [hunk 1]
```diff
@@ -53,18 +56,21 @@ export class ActionCogCreate extends CommandLineAction {
...
-        }
-        const jobData = await FileOperator.create(this.job.value).read(this.job.value);
-        const job = JSON.parse(jobData.toString()) as CogJob;
+        const jobFn = this.job?.value;
+        if (jobFn == null) throw new Error('Failed to read parameters');
+
+        const inFp = FileOperator.create(jobFn);
+        const job = JSON.parse((await inFp.read(jobFn)).toString()) as CogJob;
+        const sourceGeo = JSON.parse(
+            (await inFp.read(jobFn.replace(/job\.json/, 'source.geojson'))).toString(),
```
Comment (blacha) [inline packages/cog/src/cli/cogify/action.cog.ts:72]: `getJobPath(job, 'source.geojson')`

File: packages/cog/src/cli/cogify/action.cog.ts [hunk 2]
```diff
@@ -53,18 +56,21 @@ export class ActionCogCreate extends CommandLineAction {
...
         return quadKey;
     }
 
     async onExecute(): Promise<void> {
-        if (this.job?.value == null) {
-            throw new Error('Failed to read parameters');
-        }
-        const jobData = await FileOperator.create(this.job.value).read(this.job.value);
-        const job = JSON.parse(jobData.toString()) as CogJob;
+        const jobFn = this.job?.value;
```
Comment (blacha) [inline packages/cog/src/cli/cogify/action.cog.ts:66]: `this.job?.value!`

File: packages/cog/src/cog/cog.cutline.ts [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+function bboxIntersects(a: number[], b: number[]): boolean {
+    return modRangeOverlaps(a[0], a[2], b[0], b[2]) && modRangeOverlaps(a[1], a[3], b[1], b[3]);
+}
+
+function collectPolygons(features: Feature[] | null, bounds: Polygon): Polygon[] {
+    if (features == null || features.length == 0) return [bounds];
+    const ans: Polygon[] = [];
+
+    const addIntersect = (poly: Polygon): void => {
```
Comment (blacha) [inline packages/cog/src/cog/cog.cutline.ts:28]: Remove?

File: packages/cog/src/cog/cog.cutline.ts [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+import polygonToLine from '@turf/polygon-to-line';
+import { Feature, FeatureCollection, LineString, Polygon } from 'geojson';
+import pLimit from 'p-limit';
+import { basename } from 'path';
+import { GdalDocker } from '../gdal/gdal.docker';
+import { GdalLocal } from '../gdal/gdal.local';
+import { projectBBox, quadKeyToBbox } from '../proj';
+import { CogJob } from './cog';
+
+function modRangeOverlaps(ps: number, pe: number, qs: number, qe: number): boolean {
```
Comment (blacha) [inline packages/cog/src/cog/cog.cutline.ts:16]: Explain what this function does, what is a `mod`

File: packages/cog/src/cog/cog.cutline.ts [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+        }
+    }
+    return ans;
+}
+
+function findGeoJsonProjection(geojson: any | null): EPSG {
+    return Projection.parseEpsgString(geojson?.crs?.properties?.name ?? '') ?? EPSG.Wgs84;
+}
+
+function boundsToPolygon(bbox: number[]): Polygon {
```
Comment (blacha) [inline packages/cog/src/cog/cog.cutline.ts:69]: `toPositionPolygon`

File: packages/cog/src/cog/cog.cutline.ts [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+
+    return { type: 'Polygon', coordinates: [[topLeft, bottomLeft, bottomRight, topRight, topLeft]] };
+}
+
+function makeCutline(orig: any, polygons: Polygon[]): FeatureCollection {
+    return {
+        type: 'FeatureCollection',
+        features: polygons.map((p) => ({ type: 'Feature', properties: {}, geometry: p })),
+        name: orig.name,
+        crs: orig.crs,
```
Comment (blacha) [inline packages/cog/src/cog/cog.cutline.ts:83]: Remove all projection stuffs

File: packages/cog/src/cog/cog.cutline.ts [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+    for (const sf of bands[0].tags('SourceFilename')) {
+        ps.push(
+            queue(async () => {
+                const path = sf.textContent;
+                const tiffName = basename(path);
+                const tiffBox = sourceMap[tiffName];
+
+                if (tiffBox == null) throw new Error(`can't find ${tiffName} in vrt`);
+
+                if (!bboxIntersects(imgBox, tiffBox)) return; // image outside quadKey
```
Comment (blacha) [inline packages/cog/src/cog/cog.cutline.ts:154]: maybe add about 0.5% of buffer in all directions.

File: packages/cog/src/cog/cog.cutline.ts [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+                    if (booleanWithin(p, tiffPoly)) foundp.push(p);
+                }
+                if (foundp.length != 0) cutTifs.set(path, await cutCallback(path, makeCutline(cutline, foundp)));
+            }),
+        );
+    }
+
+    await Promise.all(ps);
+
+    for (const b of bands) {
```
Comment (blacha) [inline packages/cog/src/cog/cog.cutline.ts:191]: Have another go at this little guy.

Comment (blacha) [reply packages/cog/src/cog/cog.cutline.ts:191]: make this big guy a little guy

File: packages/cog/src/cog/cog.cutline.ts [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+    cutlinePath: string,
+    cutlineBlend: number,
+    outputFs: FileProcessor,
+    logger: LogType,
+): Promise<void> {
+    const gdal = Env.get(Env.Gdal.UseDocker, undefined) ? new GdalDocker() : new GdalLocal();
+
+    gdal.mount?.(source);
+    gdal.mount?.(target);
+    gdal.mount?.(cutlinePath);
```
Comment (blacha) [inline packages/cog/src/cog/cog.cutline.ts:247]: do we need to mount all of these,

can we just mount the temp folder that is created?

File: packages/cog/src/proj/index.ts [hunk 10]
```diff
@@ -31,3 +32,113 @@ export function guessProjection(wkt: string): EPSG | null {
...
+
+// eslint-disable-next-line @typescript-eslint/no-use-before-define
+const GoogleAREA = quadKeyToBbox('', EPSG.Google);
+
+/**
+ * Convert a bounding box to an approximate quadKey
+ * @param bbox
+ * @param epsg of the bbox
+ */
+export function bboxToQuadKey(bbox: Mercator.BBox, epsg: EPSG = EPSG.Wgs84): string {
```
Comment (blacha) [inline packages/cog/src/proj/index.ts:65]: Have a look at `@mapbox/tile-cover` see `cover.ts`

File: packages/cog/src/cog/cog.cutline.ts (hunk 5 above)
Comment (jacott) [reply packages/cog/src/cog/cog.cutline.ts:69]: Can't use wrong format

Comment (blacha) [reply packages/cog/src/cog/cog.cutline.ts:69]: what do you mean by wrong format?

File: packages/geo/src/bounds.ts [hunk 11]
```diff
@@ -66,17 +68,41 @@ export class Bounds {
         return { x: this.x, y: this.y, width: this.width, height: this.height };
     }
 
+    /**
+     * Convert to Mercator.BBox
+     */
+    public toBbox(): Mercator.BBox {
```
Comment (jacott) [inline packages/geo/src/bounds.ts:74]: Maybe use GeoJSON.BBox here instead. It is 3D as Mercator.BBox is only 2D a typescript ripple affect if I change it

File: packages/cog/src/cog/cog.cutline.ts [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+
+    return target;
+}
+
+const Cutline = {
+    buildVrt,
+    writeCutline,
+};
+
+export default Cutline;
```
Comment (blacha) [inline packages/cog/src/cog/cog.cutline.ts:205]: avoid default exports, they suck for refactoring + typescript auto imports.

just `export const Cutline ...`

**Pull request merged**